src/librustc_passes/intrinsicck.rs

```diff
@@ -119,6 +122,262 @@ impl ExprVisitor<'tcx> {
         }
         err.emit()
     }
+
+    fn is_thin_ptr_ty(&self, ty: Ty<'tcx>) -> bool {
+        if ty.is_sized(self.tcx.at(DUMMY_SP), self.param_env) {
+            return true;
+        }
+        if let ty::Foreign(..) = ty.kind {
+            return true;
+        }
+        false
+    }
+
+    fn check_asm_operand_type(
+        &self,
+        idx: usize,
+        reg: InlineAsmRegOrRegClass,
+        expr: &hir::Expr<'tcx>,
+        template: &[InlineAsmTemplatePiece],
+        tied_input: Option<(&hir::Expr<'tcx>, Option<InlineAsmType>)>,
+    ) -> Option<InlineAsmType> {
+        // Check the type against the allowed types for inline asm.
+        let ty = self.tables.expr_ty_adjusted(expr);
+        let asm_ty_isize = match self.tcx.sess.target.ptr_width {
+            16 => InlineAsmType::I16,
+            32 => InlineAsmType::I32,
+            64 => InlineAsmType::I64,
+            _ => unreachable!(),
+        };
+        let asm_ty = match ty.kind {
+            ty::Never | ty::Error => return None,
+            ty::Int(IntTy::I8) | ty::Uint(UintTy::U8) => Some(InlineAsmType::I8),
+            ty::Int(IntTy::I16) | ty::Uint(UintTy::U16) => Some(InlineAsmType::I16),
+            ty::Int(IntTy::I32) | ty::Uint(UintTy::U32) => Some(InlineAsmType::I32),
+            ty::Int(IntTy::I64) | ty::Uint(UintTy::U64) => Some(InlineAsmType::I64),
+            ty::Int(IntTy::I128) | ty::Uint(UintTy::U128) => Some(InlineAsmType::I128),
+            ty::Int(IntTy::Isize) | ty::Uint(UintTy::Usize) => Some(asm_ty_isize),
+            ty::Float(FloatTy::F32) => Some(InlineAsmType::F32),
+            ty::Float(FloatTy::F64) => Some(InlineAsmType::F64),
+            ty::FnPtr(_) => Some(asm_ty_isize),
+            ty::RawPtr(ty::TypeAndMut { ty, mutbl: _ }) if self.is_thin_ptr_ty(ty) => {
+                Some(asm_ty_isize)
+            }
+            ty::Adt(adt, substs) if adt.repr.simd() => {
```

Comment:
What about `repr(transparent)` newtypes over the allowed types?

Reply:
That was not included in the list of allowed types in the RFC. I prefer leaving it out for now, unless we start getting some demand for it.